spring-cloud-task-integration-tests/src/test/java/org/springframework/cloud/task/listener/BatchExecutionEventTests.java

```diff
@@ -0,0 +1,252 @@
+import org.springframework.cloud.task.batch.configuration.TaskBatchAutoConfiguration;
+import org.springframework.cloud.task.batch.listener.BatchEventAutoConfiguration;
+import org.springframework.context.ConfigurableApplicationContext;
+import org.springframework.context.annotation.PropertySource;
+import org.springframework.messaging.Message;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+
+/**
+ * @author Glenn Renfro
+ */
+public class BatchExecutionEventTests {
+	@ClassRule
+	public static RedisTestSupport redisTestSupport = new RedisTestSupport();
+
+	// Count for two job execution events per job
+	static CountDownLatch jobExecutionLatch = new CountDownLatch(2);
+
+	// Count for four step execution events per job
+	static CountDownLatch stepExecutionLatch = new CountDownLatch(4);
+
+	// Count for twelve item process events per job
+	static CountDownLatch itemProcessLatch = new CountDownLatch(6);
+
+	// Count for eight chunk events per job
+	static CountDownLatch chunkEventsLatch = new CountDownLatch(8);
+
+	// Count for six read events per job
+	static CountDownLatch itemReadEventsLatch = new CountDownLatch(6);
+
+	// Count for six write events per job
+	static CountDownLatch itemWriteEventsLatch = new CountDownLatch(2);
+
+	private static final String TASK_NAME = "jobEventTest";
+
+	private ConfigurableApplicationContext applicationContext;
+
+	@After
+	public void tearDown() {
+		if (applicationContext != null) {
+			applicationContext.close();
+		}
+	}
+
+	@Test
+	public void testContext() {
+		applicationContext = new SpringApplicationBuilder()
+				.sources(this.getConfigurations(BatchExecutionEventTests.ListenerBinding.class)).build()
+				.run(new String[]{ "--spring.cloud.task.closecontext.enable=false" });
+
+		assertNotNull(applicationContext.getBean("jobExecutionEventsListener"));
+		assertNotNull(applicationContext.getBean("stepExecutionEventsListener"));
+		assertNotNull(applicationContext.getBean("chunkEventsListener"));
+		assertNotNull(applicationContext.getBean("itemReadEventsListener"));
+		assertNotNull(applicationContext.getBean("itemWriteEventsListener"));
+		assertNotNull(applicationContext.getBean("itemProcessEventsListener"));
+		assertNotNull(applicationContext.getBean("skipEventsListener"));
+		assertNotNull(applicationContext.getBean(BatchEventAutoConfiguration.BatchEventsChannels.class));
+
+	}
+
+	@Test
+	public void testEventListener() throws Exception {
+		applicationContext = new SpringApplicationBuilder()
+				.sources(this.getConfigurations(BatchExecutionEventTests.ListenerBinding.class))
+				.build().run(getCommandLineParams("--spring.cloud.stream.bindings.job-execution-events.destination=foobar"));
+
+		assertTrue(jobExecutionLatch.await(1, TimeUnit.SECONDS));
+	}
```

Comment:
A couple things here:

1.  Couldn't all of these test methods be consolidated down into one test method like this:
```
@Test
public void testListeners() {
    testListener("--spring.cloud.stream.bindings.job-execution-events.destination=foobar", jobExecutionLatch, BatchExecutionEventTests.ListenerBinding.class);
    testListener("--spring.cloud.stream.bindings.step-execution-events.destination=step-execution-foobar", stepExecutionLatch, BatchExecutionEventTests. StepListenerBinding.class);
    // Insert others here
}

public void testListener(String channelBinding, CountDownLatch latch, Class<?> clazz) {
    applicationContext = new SpringApplicationBuilder()
        .sources(this.getConfigurations(clazz))
        .build().run(getCommandLineParams(channelBinding));
 
        assertTrue(latch.await(1, TimeUnit.SECONDS));
}
```

2.  Did we find out an answer on the binder method type?  If we can use a generic type like object, that removes the specific sinks for each test above.  If we can't, then we need to revisit the discussion about the aggregate channel.